src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -0,0 +1,271 @@
+from __future__ import annotations
+
+import hashlib
+import logging
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict, List, NamedTuple, Optional, Tuple, TypeVar
+
+import sentry_sdk
+from django.db import router
+from django.utils import timezone
+
+from sentry.locks import locks
+from sentry.models.artifactbundle import (
+    NULL_STRING,
+    ArtifactBundle,
+    ArtifactBundleArchive,
+    ArtifactBundleFlatFileIndex,
+    ArtifactBundleIndexingState,
+    FlatFileIndexState,
+)
+from sentry.utils import json, metrics
+from sentry.utils.db import atomic_transaction
+from sentry.utils.locking.lock import Lock
+from sentry.utils.retries import TimedRetryPolicy
+
+logger = logging.getLogger(__name__)
+
+
+# We want to keep the bundle as being indexed for 600 seconds = 10 minutes. We might need to revise this number and
+# optimize it based on the time taken to perform the indexing (on average).
+FLAT_FILE_INDEXING_CACHE_TIMEOUT = 600
+
+
+@sentry_sdk.tracing.trace
+def mark_bundle_for_flat_file_indexing(
+    artifact_bundle: ArtifactBundle,
+    project_ids: List[int],
+    release: str | None,
+    dist: str | None,
+):
+    identifiers = []
+
+    for project_id in project_ids:
+        if release:
+            identifiers.append(
+                FlatFileIdentifier(project_id, release=release, dist=dist or NULL_STRING)
+            )
+
+        identifiers.append(FlatFileIdentifier(project_id, release=NULL_STRING, dist=NULL_STRING))
+
+    # Create / Update the indexing state in the database
+    with atomic_transaction(
+        using=(
+            router.db_for_write(ArtifactBundleFlatFileIndex),
+            router.db_for_write(FlatFileIndexState),
+        )
+    ):
+        for identifier in identifiers:
+            flat_file_index, _created = ArtifactBundleFlatFileIndex.objects.get_or_create(
+                project_id=identifier.project_id,
+                release_name=identifier.release,
+                dist_name=identifier.dist,
+            )
+            FlatFileIndexState.objects.update_or_create(
+                flat_file_index=flat_file_index,
+                artifact_bundle=artifact_bundle,
+                defaults={
+                    "indexing_state": ArtifactBundleIndexingState.NOT_INDEXED.value,
```

Comment:
We need to be very careful about the semantics here, since right now we know that an identical bundle id doesn't guarantee equal content due to truncation. For this reason, re-indexing is fine, that is, we don't need to check if a specific bundle has already been indexed but it can result unexpected behavior if a certain set of uploads happens (this is because the merging of flat file indexes works under the assumption that two bundles with the same db id have identical files).

Let me explain an edge case:
- User uploads bundle with id `123` that has files `a` and `b`
- User uploads bundle with id `456` that has files `a` and `b` and `f` (notice the different id)
- User uploads bundle with id `123` that has files `c` and `d` (the id is the same since truncation can lead to conflicts).

These uploads will leave the db table `ArtifactBundle` with two entires with db id `1` and `2` where `1` points to a bundle with files `c` and `d` due to LWW semantics and `2` points to a bundle with files `a` `b` `f`. The merge algorithm works by updating the timestamp of an already existing bundle id entry and the total order of the values at a certain key (either url or debug id) are sorted by that, thus in case we end up in a situation where the user touches again `a` and `b` with a new bundle, the system will re-sort everything for `a` `b` and this new sort will use the updated timestamp for the entries with bundle `1` thus messing up ordering.

For this reason, either we have to prevent re-indexing or we have to change the merge logic to always scrap away all the entries of the bundle with the id of before. _This is done in order to keep an invariant in the index which states that for a specific bundle id, the files indexed by that bundle will only represent the most recent upload of that bundle._

Reply:
After some discussions, the high number of bundles with the same `bundle_id` but different file contents may have rather be related to `SourceBundle` not being totally deterministic.

The `bundle_id` was deterministic based on the file contents, but the zip file itself had non-deterministic timestamps which were since fixed.

If you can indeed have a bundle with the same `bundle_id` but completely different contents, then you have far more serious problems than the index being slightly inconsistent.

Reply:
What do you think about just replacing all occurrences of a bundle_id in the index with the new stuff and deleting the old?

I feel like if we keep it like it is now, we might end up with inconsistent state which might make it difficult to debug stuff.

Reply:
Done, its just a oneliner to `remove()` the bundle before we merge it into the index.